### ISSUE TITLE
Context is sometimes not complete

### ISSUE THREAD

**Issue by cdedreuille** [segmentio/analytics-react-native#541]
- `analytics-react-native` version: 2.2.0
- Integrations versions (if used):
- React Native version: 0.68
- iOS or Android or both? iOS

Following this discussion https://github.com/segmentio/analytics-react-native/issues/524 I'm trying to add a token in `context.device.token` but I still have very inconsistent behaviour. Sometimes the context will have all the data with the new token I just added but sometimes it will only have the token.

I tried to implement the new `onContextLoaded()` callback method but I might do something wrong.
Here is my code. Any chance you could tell me what I'm doing wrong?

```jsx
import messaging from '@react-native-firebase/messaging';
import {
  Plugin,
  PluginType,
  SegmentEvent,
} from '@segment/analytics-react-native';
import type { SegmentClient } from '@segment/analytics-react-native/lib/typescript/src/analytics';
import * as Device from 'expo-device';

export default class CustomerIOPlugin extends Plugin {
  type = PluginType.before;

  configure(analytics: SegmentClient) {
    this.analytics = analytics;

    this.analytics.onContextLoaded(async () => {
      const token = await messaging().getToken();
      this.analytics?.context.set({
        device: {
          token,
        },
      });
    });
  }

  execute(event: SegmentEvent) {
    return event;
  }
}
```

**Comment by oscb** [comment]
Thanks for bringing this up, this is a limitation in our current library we weren't aware of. Currently we are not deep merging the context object, so the correct way to set it should be like:
```ts
this.analytics?.context.set({
        device: {
          ...this.analytics?.context.device,
          token,
        },
      });
```

However this does not feel great. I think we should do the deepmerging where possible, specially to prevent concurrency issues. I'll work on adding that so that your code works, but feel free to use the above if you want to get unblocked for now.

**Comment by cdedreuille** [comment]
Thanks for your answer @oscb. This feels cleaner indeed but I think there's some typescript issues here, no?

<img width="781" alt="CleanShot 2022-05-16 at 21 22 08@2x" src="https://user-images.githubusercontent.com/1540635/168676084-d5319d43-2bb6-47dc-8d43-0b1e7f973ee5.png">

**Comment by cdedreuille** [comment]
I tried with this code but I still randomly get a device object with the token only.
```jsx
export default class CustomerIOPlugin extends Plugin {
  type = PluginType.before;

  configure(analytics: SegmentClient) {
    this.analytics = analytics;

    this.analytics.onContextLoaded(async () => {
      const token = await messaging().getToken();
      this.analytics?.context.set({
        device: {
          // eslint-disable-next-line @typescript-eslint/ban-ts-comment
          // @ts-ignore
          ...this.analytics?.context.device,
          token,
        },
      });
    });
  }

  execute(event: SegmentEvent) {
    return event;
  }
}
```

The only solution I could find for now is to recreate entirely the device object like this:
```jsx
export default class CustomerIOPlugin extends Plugin {
  type = PluginType.before;

  configure(analytics: SegmentClient) {
    this.analytics = analytics;

    this.analytics.onContextLoaded(async () => {
      const token = await messaging().getToken();
      this.analytics?.context.set({
        device: {
          type: Device.osName ? Device.osName : undefined,
          manufacturer: Device.manufacturer ? Device.manufacturer : undefined,
          model: Device.modelId ? Device.modelId : undefined,
          name: Device.modelName ? Device.modelName : undefined,
          token,
        },
      });
    });
  }

  execute(event: SegmentEvent) {
    return event;
  }
}
```

**Comment by 509dave16** [comment]
@cdedreuille for a short term workaround is it possible for you to change the stage at which your custom plugin operates from `PluginType.before` to `PluginType.enrichment`? The [InjectContext](https://github.com/segmentio/analytics-react-native/blob/3b61568b4fed6d2f5648b6063b5c692ffee23705/packages/core/src/plugins/Context.ts#L5) Plugin is currently operating at the `PluginType.before` stage. To ensure that it always injects the `context` before your plugin decorates it, you could just change your plugin's type to `PluginType.enrichment`.

**Comment by oscb** [comment]
@cdedreuille pardon me, the snippet I gave you was incorrect, it should be `...this.analytics?.context.get().device,`

The PR to handle deepmerging of the context just went out, will be out in the next release today/tomorrow.

By the way, the suggestion of making it an `enrichment` plugin is another good approach. It can be similar to [InjectContext](https://github.com/segmentio/analytics-react-native/blob/3b61568b4fed6d2f5648b6063b5c692ffee23705/packages/core/src/plugins/Context.ts#L5) but injecting the `token` to the events at the `enrichment` phase of the timeline. We could do a better job showing up examples like this on how to leverage the [timeline](https://github.com/segmentio/analytics-react-native#plugins--timeline-architecture) to add additional information on the events!

Anyway your original code will work as soon as we get the fix out.

**Comment by cdedreuille** [comment]
@509dave16 I tried to change the plugin type but it didn't change anything. I also tried to add this  `...this.analytics?.context.get().device` @oscb but it seems that the data always used a cache versions. Even if I updated the text it's not doing anything.

I think I'll stick to my solution of recreating the device object for now as this seems to be the most stable version but I hope this could be changed at some point in the future.
